lib/src/history/code_history_controller.dart

```diff
@@ -46,45 +52,109 @@ class CodeHistoryController {
+        lastSelection = selection;
+        _push();
+        break;
+
+      case HistoryControllerAction.newEntryAfterEdit:
+        lastCode = code;
+        lastSelection = selection;
+        _push();
+        break;
+
+      case HistoryControllerAction.setTimer:
+        _setTimer();
+        break;
     }
 
-    if (!shouldSave) {
-      if (isTextChanging) {
-        _wasTextChanged = true;
+    lastCode = code;
+    lastSelection = selection;
+
+    _removeLastRedundantSelectionOnlyChanges();
+  }
+
+  HistoryControllerAction _getAction({
+    required Code code,
+    required TextSelection selection,
+    required bool isTextChanging,
+  }) {
+    {
+      // If the line count is changed, we add record before and after edit.
+      if (code.lines.length != lastCode.lines.length) {
+        return HistoryControllerAction.newEntriesBeforeAndAfterEdit;
       }
+    }
 
-      if (_wasTextChanged) {
-        final isText1CharLonger = code.text.length == lastCode.text.length + 1;
-        final isTypingContinuous = isText1CharLonger &&
-            selection.hasMovedOneCharacterRight(lastSelection);
+    {
+      // If change is caused by casual typing, we re-set the timer.
+      final isText1CharLonger = code.text.length == lastCode.text.length + 1;
+      final isTypingContinuous = isText1CharLonger &&
+          selection.hasMovedOneCharacterRight(lastSelection);
+      if (isTypingContinuous) {
+        return HistoryControllerAction.setTimer;
+      }
+    }
 
-        if (isTypingContinuous) {
-          _setTimer();
-        } else {
-          shouldSave = true;
+    // Any other change will create record after edit.
+    return HistoryControllerAction.newEntryAfterEdit;
+  }
+
+  void _removeLastRedundantSelectionOnlyChanges() {
+    switch (stack.length) {
+      case 0:
+      case 1:
+        return;
+
+      case 2:
+        // Drop the first one if only selection has changed.
+        if (_isFullTextSame([stack[0], stack[1]])) {
+          stack.removeAt(0);
+        }
+        break;
+
+      default:
+        // Check last, last-1, last-2. Drop last-1.
+        final last = stack.last;
+        final lastMinus1 = stack[stack.length - 2];
+        final lastMinus2 = stack[stack.length - 3];
+
+        if (_isFullTextSame([last, lastMinus1, lastMinus2])) {
+          stack.removeAt(stack.length - 2);
         }
-      }
     }
+  }
 
-    if (shouldSave) {
-      _push();
+  bool _isFullTextSame(List<CodeHistoryRecord> records) {
+    final first = records.first;
+
+    for (int i = 1; i < records.length; i++) {
+      if (records[i].code.text != first.code.text) {
+        return false;
+      }
     }
 
-    lastCode = code;
-    lastSelection = selection;
+    return true;
   }
 
   void _dropRedoIfAny() {
-    stack.removeStartingAt(_currentRecordIndex + 1);
+    final startIndexToRemove = _currentRecordIndex + 1;
+    if (startIndexToRemove < stack.length) {
+      stack.removeStartingAt(startIndexToRemove);
+    }
```

Comment:
I cannot imagine an error that would require this. Maybe we need to decrement `_currentRecordIndex` in `_removeLastRedundantSelectionOnlyChanges`?
Otherwise please revert this and find the scenario when this fails, write a test, and then re-apply this.

Reply:
Yeah, decrementing fixed it.